library/std/src/io/stdio.rs

```diff
@@ -1105,7 +1113,36 @@ pub fn _print(args: fmt::Arguments<'_>) {
 #[doc(hidden)]
 #[cfg(not(test))]
 pub fn _eprint(args: fmt::Arguments<'_>) {
-    print_to(args, stderr, "stderr");
+    struct StackBuf<'a, W> {
+        buf: BorrowedBuf<'a>,
+        out: W,
+    }
+
+    impl<W: Write> Write for StackBuf<'_, W> {
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            if buf.len() >= self.buf.capacity() {
+                self.flush()?;
+                self.out.write(buf)
+            } else {
+                self.buf.unfilled().write(buf).and_then(|num| {
+                    if self.buf.unfilled().capacity() == 0 {
+                        self.flush()?;
+                    }
+                    Ok(num)
+                })
```

Comment:
If there is 1 byte left in the buffer and you attempt to write 2 bytes to it, this will cause the write to be split up in two, doesn't it? That would introduce extra places where output can get interleaved when multiple threads or programs are writing to stderr at the same time.

Reply:
Also what if the program ends up hanging/crashing in the middle of an `eprintln!()`?

Reply:
> That would introduce extra places where output can get interleaved when multiple threads or programs are writing to stderr at the same time.

No, because we hold the lock the entire time. Perhaps there's confusion with how the stdout buffer works: it keeps data around forever and tries to only write to stdout on a new line. This stderr buffer is transient and lives only for the duration of an eprintln/dbg, flushing all data to stderr before those macro calls return.

Splitting up the writes is intentional to avoid the case where someone writes in 200 byte chunks: if we tried to avoid splits, we would end up copying the 200 bytes into the buffer, then on the next 200 write out the buffer and copy those into the cleared buffer, repeat. So it would be the same as not having a buffer at all but with the performance loss of pointless copying.

> Also what if the program ends up hanging/crashing in the middle of an eprintln!()?

In both cases the eprintln will be partially or not at all written. For a panic in a fmt implementation, it goes through a different code path so it will get spotted.

Reply:
> No, because we hold the lock the entire time.

Got confused about this. My bad.

> Splitting up the writes is intentional to avoid the case where someone writes in 200 byte chunks: if we tried to avoid splits, we would end up copying the 200 bytes into the buffer, then on the next 200 write out the buffer and copy those into the cleared buffer, repeat. So it would be the same as not having a buffer at all but with the performance loss of pointless copying.

You can still flush the buffer in one write call and then write the entire 200 bytes in another write call without going through the buffer. This would avoid splitting the 200 bytes and save a copy to the buffer. (or even use `writev` to both flush the buffer and write the 200 bytes in a single syscall)

> In both cases the eprintln will be partially or not at all written.

Currently everything that is printed before the hang/crash would actually get written to stderr. With the buffer there everything that is still buffered wouldn't get written out in case of a hang/crash.

> For a panic in a fmt implementation, it goes through a different code path so it will get spotted.

That makes this less bad, but could still lead to missing information like the things that were currently in the buffer when panicking.

Reply:
> You can still flush the buffer in one write call and then write the entire 200 bytes in another write call

Yes, but then we're not using the buffer to its full potential so why bother.

> or even use writev to both flush the buffer and write the 200 bytes in a single syscall

This is brilliant! Went with this approach.

> Currently everything that is printed before the hang/crash would actually get written to stderr. With the buffer there everything that is still buffered wouldn't get written out in case of a hang/crash.

Correct and it sucks, I agree. This is basically the compromise we have to decide on. I'm feeling like the buffer isn't worth it, but at the same time it's really annoying to see all those write syscalls, so I dunno.